File: _1.16.5/src/main/java/cn/enaium/noexpensive/callback/AnvilSetOutputCallback.java
package cn.enaium.noexpensive.callback;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.item.ItemStack;

/**
 * @author Enaium
 */
public interface AnvilSetOutputCallback {
    Event<AnvilSetOutputCallback> EVENT = EventFactory.createArrayBacked(AnvilSetOutputCallback.class, (listeners) -> (output, levelCost, canTake) -> {
        for (AnvilSetOutputCallback listener : listeners) {
            listener.interact(output, levelCost, canTake);
        }
    });

    void interact(ItemStack output, int levelCost, boolean canTake);
}

File: modern/_1.21.4/src/main/kotlin/cn/enaium/noexpensive/callback/AnvilSetOutputCallback.kt
package cn.enaium.noexpensive.callback

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack

/**
 * @author Enaium
 */
fun interface AnvilSetOutputCallback {
    fun interact(output: ItemStack, levelCost: Int, canTake: Boolean, player: PlayerEntity)

    companion object {
        val EVENT: Event<AnvilSetOutputCallback> =
            EventFactory.createArrayBacked(AnvilSetOutputCallback::class.java) { listeners: Array<AnvilSetOutputCallback> ->
                AnvilSetOutputCallback { output: ItemStack, levelCost: Int, canTake: Boolean, player: PlayerEntity ->
                    for (listener in listeners) {
                        listener.interact(output, levelCost, canTake, player)
                    }
                }
            }
    }
}
